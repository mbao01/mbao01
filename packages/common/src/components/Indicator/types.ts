import { type VariantProps } from "../../libs";
import {
  getIndicatorClasses,
  getIndicatorItemClasses,
  getIndicatorPositionClasses,
} from "./constants";

type FormElements = keyof React.JSX.IntrinsicElements;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type As<P = any> =
  | FormElements
  | React.ComponentType<P>
  | React.ForwardRefExoticComponent<P>;

type AllowedElements = Pick<React.JSX.IntrinsicElements, FormElements>;

type AsProps<T extends As> = T extends FormElements
  ? AllowedElements[T]
  : T extends React.ForwardRefExoticComponent<infer P>
    ? P
    : T extends React.ComponentType<infer P>
      ? P
      : never;

export type IndicatorProps<T extends As> = AsProps<T> &
  VariantProps<typeof getIndicatorClasses> & {
    as: T;
    position?: Parameters<typeof getIndicatorPositionClasses>[0]["position"];
  };

export type IndicatorItemProps<T extends As> = AsProps<T> &
  VariantProps<typeof getIndicatorItemClasses> & {
    as: T;
  };
