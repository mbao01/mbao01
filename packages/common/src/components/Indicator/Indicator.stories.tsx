import type { Meta, StoryObj } from "@storybook/react";
import { Badge } from "../Badge";
import { Button } from "../Button";
import { Status } from "../Status";
import { Indicator } from "./Indicator";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Indicator,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Indicator>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    as: "span",
  },
  render: (args) => {
    return (
      <Indicator {...args}>
        Hello world
        <Indicator.Item {...args}>ðŸ˜ƒ</Indicator.Item>
      </Indicator>
    );
  },
};

export const IndicatorWithButtonAndStatus: Story = {
  args: {
    as: Button,
  },
  render: (args) => {
    return (
      <Indicator {...args}>
        Hello world
        <Indicator.Item as={Status} size="lg" variant="success" />
      </Indicator>
    );
  },
};

export const IndicatorWithButtonAndBadge: Story = {
  args: {
    as: Button,
  },
  render: (args) => {
    return (
      <Indicator position="start" {...args}>
        Email
        <Indicator.Item as={Badge} size="sm" variant="error">
          2
        </Indicator.Item>
      </Indicator>
    );
  },
};

export const IndicatorBottomLeft: Story = {
  args: {
    as: Button,
  },
  render: (args) => {
    return (
      <Indicator position={["start", "bottom"]} {...args}>
        Email
        <Indicator.Item as={Badge} size="sm" variant="error">
          2
        </Indicator.Item>
      </Indicator>
    );
  },
};
