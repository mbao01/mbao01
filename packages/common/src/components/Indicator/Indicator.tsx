import { Slot } from "@radix-ui/react-slot";
import type { As, IndicatorItemProps, IndicatorProps } from "./types";
import { cn } from "../../utilities";
import {
  getIndicatorClasses,
  getIndicatorItemClasses,
  getIndicatorPositionClasses,
} from "./constants";

const Indicator = <T extends As>({
  as,
  position,
  className,
  children,
  ...props
}: IndicatorProps<T>) => {
  const SlotChild = as;

  return (
    <Slot
      className={cn(getIndicatorClasses(), getIndicatorPositionClasses({ position }), className)}
      {...props}
    >
      <SlotChild>{children}</SlotChild>
    </Slot>
  );
};

const IndicatorItem = <T extends As>({
  as,
  className,
  children,
  ...props
}: IndicatorItemProps<T>) => {
  const SlotChild = as;

  return (
    <Slot className={cn(getIndicatorItemClasses(), className)} {...props}>
      <SlotChild>{children}</SlotChild>
    </Slot>
  );
};

Indicator.Item = IndicatorItem;

export { Indicator };
