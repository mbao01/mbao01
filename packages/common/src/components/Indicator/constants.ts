import { cva } from "../../libs";
import { cn } from "../../utilities";

export const getIndicatorClasses = cva("indicator");

export const getIndicatorItemClasses = cva("indicator-item");

const INDICATOR_POSITION_CLASS = {
  center: cn("indicator-center"),
  middle: cn("indicator-middle"),
  start: cn("indicator-start"),
  end: cn("indicator-end"),
  top: cn("indicator-top"),
  bottom: cn("indicator-bottom"),
};

type IndicatorPosition = keyof typeof INDICATOR_POSITION_CLASS;

export const getIndicatorPositionClasses = ({
  position = [],
}: {
  position: IndicatorPosition | IndicatorPosition[];
}) => {
  const positions = Array.isArray(position) ? position : [position];
  return positions.map((p) => INDICATOR_POSITION_CLASS[p]).join(" ");
};
