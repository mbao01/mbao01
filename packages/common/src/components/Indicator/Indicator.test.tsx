import { render, screen } from "@testing-library/react";
import { Indicator } from "./Indicator";

describe("Indicator", () => {
  it("renders a basic indicator", () => {
    const { asFragment } = render(
      <Indicator as="span">
        Main body text
        <Indicator.Item as="span">Indicator</Indicator.Item>
      </Indicator>
    );

    expect(screen.getByText("Main body text")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
