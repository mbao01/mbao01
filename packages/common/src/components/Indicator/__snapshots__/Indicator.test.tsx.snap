// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Indicator > renders a basic indicator 1`] = `
<DocumentFragment>
  <span
    class="indicator"
  >
    Main body text
    <span
      class="indicator-item"
    >
      Indicator
    </span>
  </span>
</DocumentFragment>
`;
